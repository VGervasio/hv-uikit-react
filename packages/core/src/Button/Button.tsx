import styled from "@emotion/styled";
import { variant as styledSystemVariant } from "@styled-system/variant";
import { themeVars } from "theme";

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: string;
}

const Button: React.FC<ButtonProps> = ({
  children,
  variant = "normal",
  onClick,
}) => {
  const StyledButton = styled("button")<{ variant?: string }>(
    {
      padding: 20,
      color: themeVars.colors.background,
      backgroundColor: themeVars.colors.primary,
    },
    styledSystemVariant({
      variants: {
        big: {
          fontSize: themeVars.fontSizes.title,
        },
        small: {
          fontSize: themeVars.fontSizes.normal,
        },
      },
    })
  );

  return (
    <StyledButton variant={variant} onClick={onClick}>
      {children}
    </StyledButton>
  );
};

export default Button;
