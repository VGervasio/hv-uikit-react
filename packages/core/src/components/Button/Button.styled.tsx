import styled from "@emotion/styled";
import { themeVars, themeVariant } from "theme";

const Styled = styled("button")<{ variant: string; size: string }>(
  {
    color: themeVars.colors.atmo1,
    backgroundColor: themeVars.colors.acce1,
    marginTop: themeVars.spacing[5],
    outline: "1px solid black",
  },
  themeVariant({
    variants: {
      solid: {
        fontSize: themeVars.fontSizes.base,
      },
      subtle: {
        fontSize: themeVars.fontSizes.lg,
      },
      outline: {
        fontSize: themeVars.fontSizes.sm,
      },
      ghost: {
        fontSize: themeVars.fontSizes.xs,
      },
    },
    size: {
      xs: {},
      sm: {},
      md: {},
      lg: {},
      xl: {},
    },
  })
);

export default Styled;
