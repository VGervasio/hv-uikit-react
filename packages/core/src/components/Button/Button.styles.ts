import { css, CSSObject } from "@emotion/react";
import { themeVars, themeUtils } from "theme";
import type { ButtonProps } from "./Button";

export const variantStyles: {
  [variant in NonNullable<ButtonProps["variant"]>]: CSSObject;
} = {
  primary: {
    color: themeVars.colors.atmo1,
    backgroundColor: themeVars.colors.acce2,
    ":hover": {
      backgroundColor: themeVars.colors.acce2h,
    },
  },
  secondary: {
    color: themeVars.colors.acce1,
    backgroundColor: "transparent",
    ":hover": {
      backgroundColor: themeVars.colors.atmo3,
    },
  },
};

export const sizeStyles: {
  [variant in NonNullable<ButtonProps["size"]>]: CSSObject;
} = {
  small: {
    minWidth: 80,
    lineHeight: themeVars.lineHeight.snug,
  },
  medium: {
    minWidth: 120,
    lineHeight: themeVars.lineHeight.normal,
  },
  large: {
    minWidth: 148,
    lineHeight: themeVars.lineHeight.relaxed,
  },
};

const baseStyles: CSSObject = {
  textTransform: "none",
  "&:hover,&:focus": {},
  "&:active": {},
  minWidth: "70px",
  padding: `${themeUtils.spacing(1)} ${themeUtils.spacing(
    2
  )} ${themeUtils.spacing(1)} ${themeUtils.spacing(2)}`,
  cursor: "pointer",
  minHeight: "32px",
  borderRadius: themeVars.radii.sm,
  fontSize: themeVars.fontSize.base,
  fontWeight: themeVars.fontWeight.normal,
};

export const getStyles = ({
  variant,
  size,
}: Pick<ButtonProps, "variant" | "size">): ReturnType<typeof css> => {
  const styles = {
    ...baseStyles,
    ...(variant && variantStyles[variant]),
    ...(size && sizeStyles[size]),
  };

  return css(styles);
};
