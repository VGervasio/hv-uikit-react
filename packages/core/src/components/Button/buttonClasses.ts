import { getClasses } from "@core/utils";

export interface HvButtonClasses {
  root?: string;
  startIcon?: string;
  endIcon?: string;
  focusVisible?: string;
}

const classKeys: string[] = ["root", "startIcon", "endIcon", "focusVisible"];

const buttonClasses = getClasses<HvButtonClasses>(classKeys, "HvButton");

export default buttonClasses;
