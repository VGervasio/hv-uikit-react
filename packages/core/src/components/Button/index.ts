export { default as buttonClasses } from "./buttonClasses";
export * from "./buttonClasses";
export { HvButton } from "./Button";
export type {
  HvButtonProps,
  HvButtonRadius,
  HvButtonSize,
  HvButtonVariant,
} from "./Button";
