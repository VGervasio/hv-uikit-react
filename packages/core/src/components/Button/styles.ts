import styled from "@emotion/styled";
import { themeVars, themeVariant, themeUtils } from "theme";
import { ButtonVariants } from "./Button";

export const StyledButton = styled.button<{ variant: ButtonVariants }>(
  {
    textTransform: "none",
    "&:hover,&:focus": {},
    "&:active": {},
    minWidth: "70px",
    padding: `${themeUtils.space(1)} ${themeUtils.space(2)} ${themeUtils.space(
      1
    )} ${themeUtils.space(2)}`,
    cursor: "pointer",
    minHeight: "32px",
    borderRadius: themeVars.dropdown.borderRadius,
    fontSize: themeVars.fontSizes.base,
    fontWeight: 600,
  },
  themeVariant({
    variants: {
      primary: {
        color: themeVars.colors.atmo1,
        backgroundColor: themeVars.colors.acce2,
        "&:hover": {
          backgroundColor: themeVars.colors.acce2h,
        },
      },
      primarySubtle: {
        backgroundColor: "transparent",
        border: `1px solid ${themeVars.colors.acce2}`,
        color: themeVars.colors.acce2,
        "&:hover": {
          backgroundColor: themeVars.colors.atmo3,
        },
      },
      primaryGhost: {
        color: themeVars.colors.acce2,
        backgroundColor: "transparent",
        "&:hover": {
          backgroundColor: themeVars.colors.atmo3,
        },
      },
      secondarySubtle: {
        color: themeVars.colors.acce1,
        backgroundColor: "transparent",
        border: `1px solid ${themeVars.colors.atmo4}`,
        "&:hover": {
          backgroundColor: themeVars.colors.atmo3,
          border: `1px solid ${themeVars.colors.atmo4}`,
        },
      },
      secondaryGhost: {
        color: themeVars.colors.acce1,
        backgroundColor: "transparent",
        "&:hover": {
          backgroundColor: themeVars.colors.atmo3,
        },
      },
    },
  })
);
