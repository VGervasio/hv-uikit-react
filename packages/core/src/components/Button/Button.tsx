import { CSSProperties } from "react";
import { StyledButton } from "./styles";

export type ButtonVariant =
  | "primary"
  | "primarySubtle"
  | "primaryGhost"
  | "secondarySubtle"
  | "secondaryGhost"
  // deprecated props
  | "secondary"
  | "ghost";

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  /** Use the variant prop to change the visual style of the Button. */
  variant?: ButtonVariant;
  sx?: CSSProperties;
  className?: string;
}

/**
 * Normalize the button variant. It's meant to give us some retro-compatibility with
 * the DS 3.6 API.
 *
 * @param variant the variant of the button
 * @returns       the normalized variant in DS 5 API
 */
const mapVariant = (variant: ButtonVariant): ButtonVariant => {
  if (variant === "secondary") {
    console.warn(
      "Button variant 'secondary' is deprecated. Please use 'secondarySubtle'."
    );
    return "secondarySubtle";
  }
  if (variant === "ghost") {
    console.warn(
      "Button variant 'ghost' is deprecated. Please use 'primaryGhost'."
    );
    return "primaryGhost";
  }
  return variant;
};

/**
 * Button component is used to trigger an action or event.
 */
export const Button: React.FC<ButtonProps> = ({
  children,
  variant = "primary",
  onClick,
  className,
}) => {
  return (
    <StyledButton
      className={className}
      variant={mapVariant(variant)}
      onClick={onClick}
    >
      {children}
    </StyledButton>
  );
};

export default Button;

if (process.env.NODE_ENV !== "production") {
  Button.displayName = "Button";
}
