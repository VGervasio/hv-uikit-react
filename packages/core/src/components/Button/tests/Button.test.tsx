import { describe, expect, it } from "vitest";
import { Button } from "../Button";

describe("IconComponent", () => {
  it("should match the snapshot for the given props", ({ render }) => {
    const { asFragment } = render(<Button />);
    expect(asFragment()).toMatchSnapshot();
  });
});
