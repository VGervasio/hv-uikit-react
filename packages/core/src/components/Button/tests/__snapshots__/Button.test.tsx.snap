// Vitest Snapshot v1

exports[`Button > should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="css-t79x31-StyledButton egl9i9v0"
  />
</DocumentFragment>
`;

exports[`IconComponent > should match the snapshot for the given props 1`] = `
<DocumentFragment>
  <button
    class="css-t79x31-StyledButton e138pvrm0"
  />
</DocumentFragment>
`;
